eclipse/che
# Adding k8s tolerations to workspace pods

### What does this PR do?
This PR is adding support for k8s tolerations for workspace pods.  Taints and tolerations can be used for many things, but here is why I need those:

_Node selectors & labels_: when setting a selector on a pod, it is good to look for nodes with those labels.  But it won't prevent pods without that selector to be executed on that node.

_Taints and tolerations_: if you set a taint on a node, only pods that are tolerating that taint can be executed.  The taint will protect the node from unwanted pods.

In my setup, I use a dedicated node pool in my k8s cluster for user workspace pods.  Only Che workspace pods will be executed there and nothing else.  I make that pool grow and shrink as my number of users increases or decreases.  I also use Azure SpotVMs (to lower cost) and those have a taint set by Microsoft.  So I need my pods to tolerate the SpotVM taint because otherwise, they are never scheduled to run on those.

Here is the [associated doc PR](https://github.com/eclipse/che-docs/pull/1780).

### Screenshot/screencast of this PR
In order to use tolerations, you can add them to the helm value (for now, planning to add Operator support later):
```yaml
cheWorkspacePodTolerations:
  - key: "a.node.taint"
    operator: "Equal"
    value: "aValue"
    effect: "NoExecute"
```
Helm will convert the YAML into a JSON string, which is then passed as a property to Che.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/13319

### How to test this PR?
You will need to use the helm installer since I didn't add Operator support.  In order to see the toleration in action, you will also need to add a taint on your k8s node.  See [documentation](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/).

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
